Source/WTF/wtf/Vector.h

```diff
@@ -1898,6 +1898,42 @@ Vector<typename CompactMapper<MapFunction, SourceType>::DestinationItemType> com
     return CompactMapper<MapFunction, SourceType>::compactMap(std::forward<SourceType>(source), std::forward<MapFunction>(mapFunction));
 }
 
+template<typename MapFunction, typename SourceType, typename Enable = void>
+struct FlatMapper {
+    using SourceItemType = typename CollectionInspector<SourceType>::SourceItemType;
+    using DestinationItemType = typename CollectionInspector<typename std::invoke_result<MapFunction, SourceItemType&>::type>::SourceItemType;
+
+    static Vector<DestinationItemType> flatMap(const SourceType& source, const MapFunction& mapFunction)
+    {
+        Vector<DestinationItemType> result;
+        for (auto&& item : source)
+            result.appendVector(mapFunction(item));
+        result.shrinkToFit();
+        return result;
+    }
+};
+
+template<typename MapFunction, typename SourceType>
+struct FlatMapper<MapFunction, SourceType, typename std::enable_if<std::is_rvalue_reference<SourceType&&>::value>::type> {
```

Comment:
can this use concepts instead of SFINAE?

Reply:
TBH I'm not familiar with concepts, but maybe!

Reply:
I'll try to help out!